onnxruntime/core/providers/coreml/builders/impl/reduction_op_builder.cc

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) Shukant Pal.
+// Licensed under the MIT License.
+
+#include "core/providers/common.h"
+#include "core/providers/shared/utils/utils.h"
+
+#ifdef __APPLE__
+#include "core/providers/coreml/builders/model_builder.h"
+#endif
+#include "core/providers/coreml/builders/helper.h"
+#include "core/providers/coreml/builders/op_builder_factory.h"
+
+#include "base_op_builder.h"
+
+namespace onnxruntime {
+namespace coreml {
+
+class ReductionOpBuilder : public BaseOpBuilder {
+ private:
+#ifdef __APPLE__
+  [[nodiscard]] Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                                             const logging::Logger& logger) const override;
+#endif
+};
+
+namespace {
+template <typename T>
+void AddReductionParams(T* params, std::vector<int64_t>& axes, bool keepdims) {
+  params->set_keepdims(keepdims);
+  for (auto& axis : axes) {
+    params->add_axes(axis);
+  }
+}
+} // namespace
+
+#ifdef __APPLE__
+Status ReductionOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                                                 const logging::Logger& /* logger */) const {
+  const auto& op_type(node.OpType());
+  const auto& input_defs(node.InputDefs());
+
+  NodeAttrHelper helper(node);
+  auto axes = helper.Get("axes", std::vector<int64_t>{});
+  auto keepdims = helper.Get("keepdims", false);
+
+  std::unique_ptr<COREML_SPEC::NeuralNetworkLayer> layer = CreateNNLayer(model_builder, node);
+
+  if (op_type == "ReduceSum") {
```

Comment:
there is an attribute 'noop_with_empty_axes' in the ONNX spec:
https://github.com/onnx/onnx/blob/main/docs/Operators.md#reducesum

we should either handle it or not support it (by overriding OpBuilder::IsOpSupported to check for the cases we do support). seems like it can map to the CoreML layer's reduceAll field.